### PR TITLE
Add `try_*` variants to methods on Resources that return an Option instead of panicking

### PR DESCRIPTION
**Pull request by pengowen123**
Users of the existing methods may wish to handle the error, so the new methods will return `None` in the case of a requested resource not existing.

### CODE REVIEW

File: src/res.rs [hunk 1]
```diff
@@ -249,12 +251,22 @@ impl Resources {
     pub fn fetch<T>(&self, id: usize) -> Fetch<T>
         where T: Resource
     {
-        let c = self.fetch_internal(TypeId::of::<T>(), id);
+        self.try_fetch(id).expect(RESOURCE_NOT_FOUND)
```
Comment (torkleyy) [inline src/res.rs]: Shouldn't this use `fetch_internal`?

File: src/res.rs [hunk 2]
```diff
@@ -249,12 +251,22 @@ impl Resources {
...
-        Fetch {
-            inner: c.borrow(),
-            phantom: PhantomData,
-        }
+    /// Like `fetch`, but returns an `Option` instead of panicking in the case of the resource
+    /// being accessed mutably.
+    pub fn try_fetch<T>(&self, id: usize) -> Option<Fetch<T>>
+        where T: Resource
+    {
+        let c = self.try_fetch_internal(TypeId::of::<T>(), id);
```
Comment (torkleyy) [inline src/res.rs]: Let's inline `c` here (and please to the same for the other methods, too).

File: src/res.rs [hunk 3]
```diff
@@ -263,36 +275,71 @@ impl Resources {
     pub fn fetch_mut<T>(&self, id: usize) -> FetchMut<T>
         where T: Resource
     {
-        let c = self.fetch_internal(TypeId::of::<T>(), id);
+        self.try_fetch_mut(id).expect(RESOURCE_NOT_FOUND)
```
Comment (torkleyy) [inline src/res.rs]: Same here, shouldn't it use `fetch_internal`?

File: src/res.rs (hunk 1 above)
Comment (pengowen123) [inline src/res.rs]: I wanted to avoid duplication of the code that builds the `Fetch`. Do you want me to leave it the way it was?

Comment (torkleyy) [inline src/res.rs]: I see, okay. No need to change it, right.

**Comment by pengowen123** [comment]
I have inlined the variables as requested. Some of the code may not be formatted the way you want, so you may want to run rustfmt.

File: src/res.rs [hunk 4]
```diff
@@ -263,36 +274,67 @@ impl Resources {
...
     }
 
     /// Fetches the resource with the specified type id.
     ///
     /// Please see `fetch` for details.
+    ///
+    /// # Panics
+    ///
+    /// Panics if no resource with the id exists.
     pub fn fetch_id(&self, id: TypeId, comp_id: usize) -> FetchId {
```
Comment (torkleyy) [inline src/res.rs]: Last thing: I think we should stay consistent with `fetch` then, let's use `try_fetch` here and `.expect` the value.

We should be able to get rid of `fetch_internal` then.

**Comment by pengowen123** [comment]
`fetch_internal` has been removed. If there are no other changes to make, should I squash the commits?

File: src/res.rs [hunk 5]
```diff
@@ -263,36 +274,63 @@ impl Resources {
...
     /// Fetches the resource with the specified type id mutably.
     ///
     /// Please see `fetch` for details.
+    ///
+    /// # Panics
+    ///
+    /// Panics if no resource with the id exists.
     pub fn fetch_id_mut(&self, id: TypeId, comp_id: usize) -> FetchIdMut {
-        let c = self.fetch_internal(id, comp_id);
+        let c = self.try_fetch_internal(id, comp_id).expect(RESOURCE_NOT_FOUND);
```
Comment (torkleyy) [inline src/res.rs]: Sorry for requesting changes once more, but I think you forgot to change this one to `try_fetch_id_mut().expect(...)`. And yes, please squash your commits afterwards.

**Comment by pengowen123** [comment]
I did forget that, fixed.

**Comment by kvark** [comment]
Looks like everybody is happy :)
bors r+

**Comment by kvark** [comment]
:(

**Comment by kvark** [comment]
bors - y u no like me?

**Comment by notriddle** [comment]
Apparently, @torkleyy is the only current reviewer for the shred repo.

**Comment by torkleyy** [comment]
I'm sure I changed it last time..

**Comment by notriddle** [comment]
[Here it is, with just you in the list](https://bors-app.herokuapp.com/repositories/128/settings)

**Comment by torkleyy** [comment]
Okay, but now it should work. Thanks.

**Pull request merged**